### PR TITLE
Use full import path as 'type' in serialization. 

### PR DESCRIPTION
**Pull request by masci**
fixes: https://github.com/deepset-ai/haystack/issues/6186

## Changes

- Use full import path of the component as `type` in its dict representation
- Try to import the full path when deserializing a component

Before:
```py
pipeline_dict = {
    "components": {
        "greeter": {
            "type": "Greet",
            "init_parameters": {
                "message": "\nGreeting component says: Hi! The value is {value}\n",
                "log_level": "INFO",
            },
        }
    },
   ...
}
```

After:
```py
pipeline_dict = {
    "components": {
        "greeter": {
            "type": "sample_components.greet.Greet",  # <-- Now we use the full import path to represent the type
            "init_parameters": {
                "message": "\nGreeting component says: Hi! The value is {value}\n",
                "log_level": "INFO",
            },
        }
    },
   ...
}
```

### CODE REVIEW

File: test/sample_components/test_add_value.py [hunk 1]
```diff
@@ -5,30 +5,6 @@
...
-    component = component_from_dict(AddFixedValue, data)
-    assert component.add == 1
-
-
-def test_from_dict_with_custom_add_value():
-    data = {"type": "AddFixedValue", "init_parameters": {"add": 100}}
-    component = component_from_dict(AddFixedValue, data)
-    assert component.add == 100
-
-
```
Comment (masci) [inline test/sample_components/test_add_value.py:31]: Removed serialization tests for components that use the default to_dict method

File: test/sample_components/test_concatenate.py [hunk 2]
```diff
@@ -5,18 +5,6 @@
...
-    res = component_to_dict(component)
-    assert res == {"type": "Concatenate", "init_parameters": {}}
-
-
-def test_from_dict():
-    data = {"type": "Concatenate", "init_parameters": {}}
-    component = component_from_dict(Concatenate, data)
-    assert component
-
-
```
Comment (masci) [inline test/sample_components/test_concatenate.py:19]: Removed serialization tests for components that use the default to_dict method

File: test/sample_components/test_double.py [hunk 3]
```diff
@@ -6,18 +6,6 @@
...
-    res = component_to_dict(component)
-    assert res == {"type": "Double", "init_parameters": {}}
-
-
-def test_from_dict():
-    data = {"type": "Double", "init_parameters": {}}
-    component = component_from_dict(Double, data)
-    assert component
-
-
```
Comment (masci) [inline test/sample_components/test_double.py:20]: Removed serialization tests for components that use the default to_dict method

File: test/sample_components/test_greet.py [hunk 4]
```diff
@@ -7,50 +7,6 @@
...
-def test_from_with_custom_parameters():
-    data = {
-        "type": "Greet",
-        "init_parameters": {"message": "My message", "log_level": "ERROR"},
-    }
-    component = component_from_dict(Greet, data)
-    assert component.message == "My message"
-    assert component.log_level == "ERROR"
-
-
```
Comment (masci) [inline test/sample_components/test_greet.py:53]: Removed serialization tests for components that use the default to_dict method

File: test/sample_components/test_parity.py [hunk 5]
```diff
@@ -5,18 +5,6 @@
...
-    res = component_to_dict(component)
-    assert res == {"type": "Parity", "init_parameters": {}}
-
-
-def test_from_dict():
-    data = {"type": "Parity", "init_parameters": {}}
-    component = component_from_dict(Parity, data)
-    assert component
-
-
```
Comment (masci) [inline test/sample_components/test_parity.py:19]: Removed serialization tests for components that use the default to_dict method

File: test/sample_components/test_remainder.py [hunk 6]
```diff
@@ -7,30 +7,6 @@
...
-    component = component_from_dict(Remainder, data)
-    assert component.divisor == 3
-
-
-def test_from_dict_with_custom_divisor_value():
-    data = {"type": "Remainder", "init_parameters": {"divisor": 100}}
-    component = component_from_dict(Remainder, data)
-    assert component.divisor == 100
-
-
```
Comment (masci) [inline test/sample_components/test_remainder.py:33]: Removed serialization tests for components that use the default to_dict method

File: test/sample_components/test_repeat.py [hunk 7]
```diff
@@ -5,18 +5,6 @@
...
-    res = component_to_dict(component)
-    assert res == {"type": "Repeat", "init_parameters": {"outputs": ["first", "second"]}}
-
-
-def test_from_dict():
-    data = {"type": "Repeat", "init_parameters": {"outputs": ["first", "second"]}}
-    component = component_from_dict(Repeat, data)
-    assert component.outputs == ["first", "second"]
-
-
```
Comment (masci) [inline test/sample_components/test_repeat.py:19]: Removed serialization tests for components that use the default to_dict method

File: test/sample_components/test_subtract.py [hunk 8]
```diff
@@ -5,18 +5,6 @@
...
-    res = component_to_dict(component)
-    assert res == {"type": "Subtract", "init_parameters": {}}
-
-
-def test_from_dict():
-    data = {"type": "Subtract", "init_parameters": {}}
-    component = component_from_dict(Subtract, data)
-    assert component
-
-
```
Comment (masci) [inline test/sample_components/test_subtract.py:19]: Removed serialization tests for components that use the default to_dict method

File: test/sample_components/test_sum.py [hunk 9]
```diff
@@ -6,17 +6,6 @@
...
-    component = Sum()
-    res = component_to_dict(component)
-    assert res == {"type": "Sum", "init_parameters": {}}
-
-
-def test_from_dict():
-    data = {"type": "Sum", "init_parameters": {}}
-    component = component_from_dict(Sum, data)
-
-
```
Comment (masci) [inline test/sample_components/test_sum.py:19]: Removed serialization tests for components that use the default to_dict method

File: test/sample_components/test_threshold.py [hunk 10]
```diff
@@ -5,30 +5,6 @@
...
-    component = component_from_dict(Threshold, data)
-    assert component.threshold == 10
-
-
-def test_from_dict_with_custom_threshold_value():
-    data = {"type": "Threshold", "init_parameters": {"threshold": 100}}
-    component = component_from_dict(Threshold, data)
-    assert component.threshold == 100
-
-
```
Comment (masci) [inline test/sample_components/test_threshold.py:31]: Removed serialization tests for components that use the default to_dict method

File: canals/component/component.py [hunk 11]
```diff
@@ -265,15 +265,16 @@ def copy_class_namespace(namespace):
         class_ = new_class(class_.__name__, class_.__bases__, {"metaclass": ComponentMeta}, copy_class_namespace)
 
         # Save the component in the class registry (for deserialization)
-        if class_.__name__ in self.registry:
+        class_path = f"{class_.__module__}.{class_.__name__}"
```
Comment (masci) [inline canals/component/component.py:268]: Also use the full path as the key in `component.registry`

File: canals/pipeline/pipeline.py [hunk 12]
```diff
@@ -152,8 +153,18 @@ def from_dict(cls: Type[T], data: Dict[str, Any], **kwargs) -> T:
...
-                    raise PipelineError(f"Component '{component_data['type']}' not imported.")
+                    try:
+                        # Import the module first...
+                        module, _ = component_data["type"].rsplit(".", 1)
+                        importlib.import_module(module)
+                        # ...then try again
+                        if component_data["type"] not in component.registry:
+                            raise PipelineError()
+                    except (ImportError, PipelineError) as e:
+                        raise PipelineError(f"Component '{component_data['type']}' not imported.") from e
```
Comment (masci) [inline canals/pipeline/pipeline.py:166]: This is the core logic: in `from_dict` we try to import the module containing the type of the components listed in the dictionary.

**Review by anakin87 — APPROVED** [review]
I have had a look and the new logic looks fine to me...

Since I'm not (yet) a canals expert, feel free to ask for a review from more experienced people as well (if you feel it would be helpful).

File: canals/pipeline/pipeline.py [hunk 13]
```diff
@@ -152,8 +153,18 @@ def from_dict(cls: Type[T], data: Dict[str, Any], **kwargs) -> T:
             else:
                 if "type" not in component_data:
                     raise PipelineError(f"Missing 'type' in component '{name}'")
+
                 if component_data["type"] not in component.registry:
-                    raise PipelineError(f"Component '{component_data['type']}' not imported.")
+                    try:
+                        # Import the module first...
+                        module, _ = component_data["type"].rsplit(".", 1)
+                        importlib.import_module(module)
```
Comment (ZanSara) [inline canals/pipeline/pipeline.py:161]: Nitpick: I would add a debug log here
```suggestion
                        logger.debug("Trying to import %s", module)
                        importlib.import_module(module)
```

File: canals/pipeline/pipeline.py [hunk 14]
```diff
@@ -152,8 +153,18 @@ def from_dict(cls: Type[T], data: Dict[str, Any], **kwargs) -> T:
...
+                        # ...then try again
+                        if component_data["type"] not in component.registry:
+                            raise PipelineError()
```
Comment (ZanSara) [inline canals/pipeline/pipeline.py:164]: Nitpick: even if it's close to impossible, I'd add a message to this error asking to open a bug report to Canals.

File: test/test_serialization.py [hunk 15]
```diff
@@ -58,3 +60,32 @@ def test_default_component_from_dict_unregistered_component(request):
...
+    }
+
+    # remove the target component from the registry if already there
+    component.registry.pop("sample_components.greet.Greet", None)
+    # remove the module from sys.modules if already there
+    sys.modules.pop("sample_components.greet", None)
+
+    p = Pipeline.from_dict(pipeline_dict)
+
+    from sample_components.greet import Greet
```
Comment (ZanSara) [inline test/test_serialization.py:88]: Nitpick: `Greet` should be already imported (if I understood the code above). If so, we may add `if TYPE_CHECKING:` here to make it available only to IDEs and not to the test.
```suggestion
    if TYPE_CHECKING:
        from sample_components.greet import Greet
```

**Review by ZanSara — APPROVED** [review]
Left only a couple of optional nitpicks

File: test/test_serialization.py (hunk 15 above)
Comment (masci) [reply test/test_serialization.py:88]: That's not for type checking, I actually need the symbol `Greet` to compare with the output of `get_component`

**Pull request merged**